src/testing/vortex/faulty_network.zig

```diff
@@ -0,0 +1,583 @@
+//! or if there's an error or EOF, in which case it also closes the parent connection.
+//!
+//! When the `Faults` struct is populated with non-null configurations, the pipes inject
+//! corresponding faults according to the probabilities.
+//!
+//! NOTE: The pipe is not yet message-aware (and perhaps shouldn't be?), which means that we deal
+//! with whatever chunks of bytes we receieve immediately, instead of collecting a buffer with a
+//! full message before piping it through.
+const std = @import("std");
+const arbitrary = @import("./arbitrary.zig");
+const IO = @import("../../io.zig").IO;
+const constants = @import("./constants.zig");
+
+const assert = std.debug.assert;
+const log = std.log.scoped(.faulty_network);
+
+const Faults = struct {
+    const Delay = struct {
+        time_ms: u32,
+        jitter_ms: u32,
+    };
+
+    const Lose = struct {
+        percentage: u8,
+    };
+
+    const Corrupt = struct {
+        percentage: u8,
+    };
+
+    delay: ?Delay = null,
+    lose: ?Lose = null,
+    corrupt: ?Corrupt = null,
+
+    // Others not implemented: duplication, reordering, rate
+
+    pub fn heal(faults: *Faults) void {
+        faults.delay = null;
+        faults.lose = null;
+        faults.corrupt = null;
+    }
+
+    pub fn is_healed(faults: *const Faults) bool {
+        return faults.delay == null and faults.lose == null and faults.corrupt == null;
+    }
+};
+
+pub const Mapping = struct {
+    remote: std.net.Address,
+    origin: std.net.Address,
+};
+
+const Pipe = struct {
+    io: *IO,
+    connection: *Connection,
+    input: std.posix.socket_t = IO.INVALID_SOCKET,
+    output: std.posix.socket_t = IO.INVALID_SOCKET,
+    buffer: [constants.message_size_max]u8 = undefined,
+    send_inflight: bool = false,
+    recv_inflight: bool = false,
+    recv_count: usize = 0,
+    send_count: usize = 0,
+
+    recv_completion: IO.Completion = undefined,
+    send_completion: IO.Completion = undefined,
+
+    fn open(
+        pipe: *Pipe,
+        input: std.posix.socket_t,
+        output: std.posix.socket_t,
+    ) void {
+        assert(pipe.connection.state == .proxying);
+        assert(input != IO.INVALID_SOCKET);
+        assert(output != IO.INVALID_SOCKET);
+
+        pipe.input = input;
+        pipe.output = output;
+        pipe.recv_count = 0;
+        pipe.send_count = 0;
+
+        // Kick off the recv/send loop.
+        pipe.recv();
+    }
+
+    fn has_inflight_operations(pipe: *const Pipe) bool {
+        return pipe.recv_inflight or pipe.send_inflight;
+    }
+
+    fn recv(pipe: *Pipe) void {
+        assert(!pipe.recv_inflight);
+        assert(pipe.connection.state == .proxying);
+
+        pipe.recv_count = 0;
+        pipe.send_count = 0;
+        pipe.recv_inflight = true;
+
+        pipe.connection.io.recv(
+            *Pipe,
+            pipe,
+            on_recv,
+            &pipe.recv_completion,
+            pipe.input,
+            pipe.buffer[0..],
+        );
```

Comment:
Ah, I guess we don't need to handle partial recv as anything special, because we are a proxy - if we recv partially and send all that on, that's OK. Can you make a comment about that? (I was looking for recv all logic)